### PR TITLE
Implement zip with lambda for two arrays

### PR DESCRIPTION
**Pull request by rongrong**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                null);
+    }
+
+    private static Type zipReturnType(Type... types)
+    {
+        return new ArrayType(new RowType(list(types), Optional.empty()));
+    }
+
+    @SafeVarargs
+    private static <T> List<T> list(T... a)
```
Comment (geraint0923) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java]: I think it is better to use `ImmutableList.of(...)`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    public static final ZipTransformFunction ZIP_TRANSFORM_FUNCTION = new ZipTransformFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ZipTransformFunction.class, "transform", Type.class, Type.class, Type.class, Block.class, Block.class, MethodHandle.class);
+
+    private ZipTransformFunction()
+    {
+        super(new Signature(
+                "zip",
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("T"), typeVariable("U"), typeVariable("V")),
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java]: I would use`T`, `U`, `R`. R stands for result. S stands for state.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        extends SqlScalarFunction
+{
+    public static final ZipTransformFunction ZIP_TRANSFORM_FUNCTION = new ZipTransformFunction();
+
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ZipTransformFunction.class, "transform", Type.class, Type.class, Type.class, Block.class, Block.class, MethodHandle.class);
+
+    private ZipTransformFunction()
+    {
+        super(new Signature(
+                "zip",
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java]: As we discussed and agreed, we will call this merge. TODO

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java (hunk 1 above)
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java]: Use `ImmutableList.of` whenever possible. When `null` element is present, use `Arrays.asList`. It's not necessary to introduce this `list` function.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    @Test
+    public void testWithNull()
+            throws Exception
+    {
+        assertFunction("zip(CAST(NULL AS ARRAY(UNKNOWN)), ARRAY[], (x, y) -> (y, x))",
+                zipReturnType(UNKNOWN, UNKNOWN),
+                null);
+    }
+
+    private static Type zipReturnType(Type... types)
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipTransformFunction.java]: I would inline this function. 
* The name as is does not properly convey what this function does.
* A proper name for this function is hard.
* The function appears to be simple enough to just inline.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            Object output;
+            try {
+                output = function.invoke(left, right);
+            }
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+            writeNativeValue(outputType, resultBuilder, output);
+        }
+        if (leftPositionCount > shortest) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipTransformFunction.java]: IIRC, we discusses and agreed that the function will fail if the two arrays are not of the same length.

**Comment by haozhun** [comment]
Back to you, @rongrong

**Comment by martint** [comment]
Given the discussion about why we can't use `zip` or `merge` to name this function, I'd go with `zip_with`, as in "zip these arrays with this lambda expression"

**Comment by rongrong** [comment]
Addressed comments. Please review @haozhun

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        assertInvalidFunction("zip_with(ARRAY[1], ARRAY['a', 'b'], (x, y) -> (y, x))", "Two arrays have different length");
+        assertInvalidFunction("zip_with(ARRAY[NULL, 2], ARRAY['a'], (x, y) -> (y, x))", "Two arrays have different length");
+        assertInvalidFunction("zip_with(ARRAY[1, NULL], ARRAY[3, 2, 1], (x, y) -> x + y)", "Two arrays have different length");
+    }
+
+    @Test
+    public void testWithNull()
+            throws Exception
+    {
+        assertFunction("zip_with(CAST(NULL AS ARRAY(UNKNOWN)), ARRAY[], (x, y) -> (y, x))",
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java]: Please test cases where
* the array contains null, but the lambda returns non-null always
* the array contains null, the lambda returns null when one of the input is null
* the array doesn't contain null, the lambda returns null for some non-null input

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "merge two arrays, element-wise, into a single array of rows, and apply lambda to each row";
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java]: merge two arrays, element-wise, into a single array using the lambda function

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(false, false, false),
+                METHOD_HANDLE.bindTo(leftArrayType).bindTo(rightArrayType).bindTo(outputType),
+                isDeterministic());
+    }
+
+    public static Block zipWith(Type leftArrayType, Type rightArrayType, Type outputType, Block leftBlock, Block rightBlock, MethodHandle function)
+    {
+        checkCondition(leftBlock.getPositionCount() == rightBlock.getPositionCount(), INVALID_FUNCTION_ARGUMENT, "Two arrays have different length");
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java]: `Arrays must have the same length`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        return "merge two arrays, element-wise, into a single array of rows, and apply lambda to each row";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type leftArrayType = boundVariables.getTypeVariable("T");
+        Type rightArrayType = boundVariables.getTypeVariable("U");
+        Type outputType = boundVariables.getTypeVariable("R");
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java]: Call them `leftElementType`, `rightElementType`, and `outputElementType` to make it apparent that the type has stripped away the top-level `array<>`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(false, false, false),
+                METHOD_HANDLE.bindTo(leftArrayType).bindTo(rightArrayType).bindTo(outputType),
+                isDeterministic());
+    }
+
+    public static Block zipWith(Type leftArrayType, Type rightArrayType, Type outputType, Block leftBlock, Block rightBlock, MethodHandle function)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ZipWithFunction.java]: `leftElementType`, `rightElementType`, and `outputElementType`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.facebook.presto.spi.type.VarcharType.createVarcharType;
+import static com.facebook.presto.type.UnknownType.UNKNOWN;
+import static java.util.Arrays.asList;
+
+public class TestZipWithFunction
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testSameLength()
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java]: Make sure types with all 5 java types are covered.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        assertFunction("zip_with(ARRAY[1, 2], ARRAY['a', 'b'], (x, y) -> (y, x))",
+                new ArrayType(new RowType(ImmutableList.of(createVarcharType(1), INTEGER), Optional.empty())),
+                ImmutableList.of(ImmutableList.of("a", 1), ImmutableList.of("b", 2)));
+
+        assertFunction("zip_with(ARRAY[1, 2], ARRAY[CAST('a' AS VARCHAR), CAST('b' AS VARCHAR)], (x, y) -> (y, x))",
+                new ArrayType(new RowType(ImmutableList.of(VARCHAR, INTEGER), Optional.empty())),
+                ImmutableList.of(ImmutableList.of("a", 1), ImmutableList.of("b", 2)));
+
+        assertFunction("zip_with(ARRAY[], ARRAY[], (x, y) -> (y, x))",
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java]: I would move this test to the beginning of this method.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+                new ArrayType(new RowType(ImmutableList.of(UNKNOWN, UNKNOWN), Optional.empty())),
+                ImmutableList.of());
+
+        assertFunction("zip_with(ARRAY[NULL], ARRAY[NULL], (x, y) -> (y, x))",
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestZipWithFunction.java]: I would move this to `testWithNull`

**Comment by haozhun** [comment]
Change commit message to "Add zip_with function"

**Comment by rongrong** [comment]
Addressed comments and added more tests.

**Pull request closed without merging**